File: lua/test/core_entity_test.lua
-- Core entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("spacex-rest_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("CoreEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:Core(nil)
    assert.is_not_nil(ent)
  end)

  it("should run basic flow", function()
    local setup = core_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"list", "load"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "core." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set SPACEXREST_TEST_CORE_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- Bootstrap entity data from existing test data.
    local core_ref01_data_raw = vs.items(helpers.to_map(
      vs.getpath(setup.data, "existing.core")))
    local core_ref01_data = nil
    if #core_ref01_data_raw > 0 then
      core_ref01_data = helpers.to_map(core_ref01_data_raw[1][2])
    end

    -- LIST
    local core_ref01_ent = client:Core(nil)
    local core_ref01_match = {}

    local core_ref01_list_result, err = core_ref01_ent:list(core_ref01_match, nil)
    assert.is_nil(err)
    assert.is_table(core_ref01_list_result)

    -- LOAD
    local core_ref01_match_dt0 = {
      id = core_ref01_data["id"],
    }
    local core_ref01_data_dt0_loaded, err = core_ref01_ent:load(core_ref01_match_dt0, nil)
    assert.is_nil(err)
    local core_ref01_data_dt0_load_result = helpers.to_map(core_ref01_data_dt0_loaded)
    assert.is_not_nil(core_ref01_data_dt0_load_result)
    assert.are.equal(core_ref01_data_dt0_load_result["id"], core_ref01_data["id"])

  end)
end)

function core_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/core/CoreTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read core test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "core01", "core02", "core03" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("SPACEXREST_TEST_CORE_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["SPACEXREST_TEST_CORE_ENTID"] = idmap,
    ["SPACEXREST_TEST_LIVE"] = "FALSE",
    ["SPACEXREST_TEST_EXPLAIN"] = "FALSE",
  })

  local idmap_resolved = helpers.to_map(
    env["SPACEXREST_TEST_CORE_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end

  if env["SPACEXREST_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      {
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["SPACEXREST_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["SPACEXREST_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end

File: php/test/CoreEntityTest.php
<?php
declare(strict_types=1);

// Core entity test

require_once __DIR__ . '/../spacexrest_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class CoreEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = SpacexRestSDK::test(null, null);
        $ent = $testsdk->Core(null);
        $this->assertNotNull($ent);
    }

    public function test_basic_flow(): void
    {
        $setup = core_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["list", "load"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "core." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set SPACEXREST_TEST_CORE_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // Bootstrap entity data from existing test data.
        $core_ref01_data_raw = Vs::items(Helpers::to_map(
            Vs::getpath($setup["data"], "existing.core")));
        $core_ref01_data = null;
        if (count($core_ref01_data_raw) > 0) {
            $core_ref01_data = Helpers::to_map($core_ref01_data_raw[0][1]);
        }

        // LIST
        $core_ref01_ent = $client->Core(null);
        $core_ref01_match = [];

        $core_ref01_list_result = $core_ref01_ent->list($core_ref01_match, null);
        $this->assertIsArray($core_ref01_list_result);

        // LOAD
        $core_ref01_match_dt0 = [
            "id" => $core_ref01_data["id"],
        ];
        $core_ref01_data_dt0_loaded = $core_ref01_ent->load($core_ref01_match_dt0, null);
        $core_ref01_data_dt0_load_result = Helpers::to_map($core_ref01_data_dt0_loaded);
        $this->assertNotNull($core_ref01_data_dt0_load_result);
        $this->assertEquals($core_ref01_data_dt0_load_result["id"], $core_ref01_data["id"]);

    }
}

function core_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/core/CoreTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = SpacexRestSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["core01", "core02", "core03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("SPACEXREST_TEST_CORE_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "SPACEXREST_TEST_CORE_ENTID" => $idmap,
        "SPACEXREST_TEST_LIVE" => "FALSE",
        "SPACEXREST_TEST_EXPLAIN" => "FALSE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["SPACEXREST_TEST_CORE_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["SPACEXREST_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
            ],
            $extra ?? [],
        ]);
        $client = new SpacexRestSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["SPACEXREST_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["SPACEXREST_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}
